lib/attrsets.nix

```
@@ -1726,6 +1728,75 @@ rec {
     if path == [] then "<root attribute path>"
     else concatMapStringsSep "." escapeNixIdentifier path;
 
+  /**
+    Turns a list of strings into a human-readable description of those
+    strings represented as an attribute path. The result of this function is
+    not intended to be machine-readable.
+    Create a new attribute set with `value` set at the nested attribute location specified in `attrPath`.
+
+    # Inputs
+
+    `context` (String)
+
+    : The error context in case a tag doesn't have a handler or the value is not valid.
+
+    `value` (AttrsOf Any)
+
+    : The value to match.
+
+    `handlers` (AttrsOf (Any -> Any))
+
+    : An attribute set containing a handler function for each case.
+
+    # Type
+
+    ```
+    matchAttrTag :: String -> AttrsOf Any -> AttrsOf (Any -> Any) -> Any
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.attrsets.matchAttrTag` usage example
+
+    ```nix
+    matchAttrTag "test" { n = 10; } {
+      n = number: "It's the number ${toString number}";
+      s = str: "It's the string ${str}";
+    }
+    => "It's the number 10"
+
+    matchAttrTag "test" { s = "Paul"; } {
+      n = number: "It's the number ${toString number}";
+      s = str: "It's the string ${str}";
+    }
+    => "It's the string Paul"
+
+    matchAttrTag "test" { unknown = null; } {
+      n = number: "It's the number ${toString number}";
+      s = str: "It's the string ${str}";
+    }
+    => error: test: No handler for tag "unknown"
+
+    matchAttrTag "test" { n = 10; s = "Paul"; } {
+      n = number: "It's the number ${toString number}";
+      s = str: "It's the string ${str}";
+    }
+    => error: test: Value has not exactly one tag: [ "n" "s" ]
+    ```
+
+    :::
+  */
+  matchAttrTag =
+    context: value: handlers:
```

```suggestion
    context: handlers: value:
```

Often, the `handlers` are more part of the program than part of the data, so arranging them like this lets us take advantage of currying, in things like `map`, pipes, or eta-reduced function definitions.

---

I think the most common case is to _not_ curry tbh, and you can still use `flip` if you need it. With the `value: handlers:` order we're also mirroring how matching looks like in many other languages:

```rust
match value {
  Foo { .. } => 0
  Bar { .. } => 1
}
```

```haskell
case value of
  Foo { .. } -> 0
  Bar { .. } -> 1
```

So I'd really prefer leaving it like that

---

The annoying thing is that if you do want to lib.flip, you have to write it like `flip (matchAttrTag "context") [...]` which looks a bit funny to me

---

I guess if we go for https://github.com/NixOS/nixpkgs/pull/304328#discussion_r1598758011 this won't be a problem anymore :D

---

**Pipe friendly**:

```nix
map (matchAttrTag "context") myList
```
and
```nix
value |> matchAttrTag "context" {
  ...
}
```

**Not pipe friendly**:

```nix
map (flip (matchAttrTag "context")) myList
```
and
```nix
matchAttrTag "context" value {
  ...
}
```

---

Also, fwiw, avoiding `flip` avoids a runtime cost.